insert_string_tpl.h

```diff
@@ -22,33 +22,58 @@
  *
  */
 
+#ifndef HASH_CALC_OFFSET
+#  define HASH_CALC_OFFSET 0 
+#endif
+#ifndef HASH_CALC_MASK
+#  define HASH_CALC_MASK HASH_MASK
+#endif
+#ifndef HASH_CALC_READ
+#  ifdef UNALIGNED_OK
+#    define HASH_CALC_READ \
+        val = *(uint32_t *)(strstart);
+#  else
+#    define HASH_CALC_READ \
+        val  = ((uint32_t)(strstart[0])); \
+        val |= ((uint32_t)(strstart[1]) << 8); \
+        val |= ((uint32_t)(strstart[2]) << 16); \
+        val |= ((uint32_t)(strstart[3]) << 24);
+#  endif
+#endif
+
+/* ===========================================================================
+ * Update a hash value with the given input byte
+ * IN  assertion: all calls to to UPDATE_HASH are made with consecutive
+ *    input characters, so that a running hash key can be computed from the
+ *    previous key instead of complete recalculation each time.
+ */
+Z_INTERNAL uint32_t UPDATE_HASH(deflate_state *const s, uint32_t h, uint32_t val) {
+    HASH_CALC(s, h, val);
+    return h & HASH_CALC_MASK;
+}
+
 /* ===========================================================================
  * Quick insert string str in the dictionary and set match_head to the previous head
  * of the hash chain (the most recent string with same hash key). Return
  * the previous length of the hash chain.
  */
-Z_INTERNAL Pos QUICK_INSERT_STRING(deflate_state *const s, const uint32_t str) {
+Z_INTERNAL Pos QUICK_INSERT_STRING(deflate_state *const s, uint32_t str) {
     Pos head;
-    uint8_t *strstart = s->window + str;
-    uint32_t val, hm, h = 0;
-
-#ifdef UNALIGNED_OK
-    val = *(uint32_t *)(strstart);
-#else
-    val  = ((uint32_t)(strstart[0]));
-    val |= ((uint32_t)(strstart[1]) << 8);
-    val |= ((uint32_t)(strstart[2]) << 16);
-    val |= ((uint32_t)(strstart[3]) << 24);
-#endif
+    uint8_t *strstart = s->window + str + HASH_CALC_OFFSET;
+    uint32_t val, hm;
 
-    UPDATE_HASH(s, h, val);
-    hm = h & HASH_MASK;
+    HASH_CALC_VAR_INIT;
```

Comment:
@Dead2 and @mtl1979 any thoughts on this function? I am using a lot of #predefines. In one way it is nice in that `HASH_CALC_READ` has a different value depending on if alignment is enabled.

In another way, with the rolling sum it might be harder to read. I wonder if insert_string_roll.h should not use the template header. But then again all the other hash table logic would be the same.. Thoughts?

Reply:
I'm not sure about the name... The name doesn't clearly say what the contents do.

Reply:
I’m talking about using more macros in these insert string functions such as QUICK_INSERT_STRING for rolling hash sum.

Reply:
What I'm trying to say is that too many macros make it hard to follow what the code actually does.

Reply:
Actually looking back on it I think I am ok with it.